src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationPartFactory.cs

```diff
@@ -27,5 +34,57 @@ public abstract class ApplicationPartFactory
         /// The context name. By default, value of this parameter is <see cref="DefaultContextName"/>.
         /// </param>
         public abstract IEnumerable<ApplicationPart> GetApplicationParts(Assembly assembly, string context);
+
+        /// <summary>
+        /// Gets the <see cref="ApplicationPartFactory"/> for the specified assembly.
+        /// <para>
+        /// An assembly may specify an <see cref="ApplicationPartFactory"/> using <see cref="ProvideApplicationPartFactoryAttribute"/>.
+        /// Otherwise, <see cref="ApplicationPartFactory.Default"/> is used.
+        /// </para>
+        /// </summary>
+        /// <param name="assembly">The <see cref="Assembly"/>.</param>
+        /// <returns>An instance of <see cref="ApplicationPartFactory"/>.</returns>
+        public static ApplicationPartFactory GetApplicationPartFactory(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            var provideAttribute = assembly.GetCustomAttribute<ProvideApplicationPartFactoryAttribute>();
+            if (provideAttribute == null)
+            {
+                return ApplicationPartFactory.Default;
+            }
+
+            var type = provideAttribute.ApplicationPartFactoryType;
+            if (type == null)
+            {
+                type = Type.GetType(provideAttribute.ApplicationPartFactoryTypeName, throwOnError: true);
+            }
+
+            if (!typeof(ApplicationPartFactory).IsAssignableFrom(type))
+            {
+                throw new InvalidOperationException(Resources.FormatApplicationPartFactory_InvalidFactoryType(
+                    type,
+                    nameof(ProvideApplicationPartFactoryAttribute),
+                    typeof(ApplicationPartFactory)));
+            }
+
+            return (ApplicationPartFactory)Activator.CreateInstance(type);
+        }
+
+        private class DefaultApplicationPartFactory : ApplicationPartFactory
```

Comment:
Consider just making this public and top level. It seems fundamental

Reply:
There's a way to get to it via `ApplicationPartFactory.Default`. Doesn't seem like there'd be much value in letting users new it up.